Don't trim highlight snippet if number_of_fragments is 0 #101803

yashdamani opened the pull request:
Don't trim highlight snippet if number_of_fragments is 0

Closes  #101803

yashdamani committed:
added fix for inconsistent text trimming in Unified Highlighter #101803

yashdamani committed:
added changelog file

AlexHelmutSonntag commented on server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java:
```diff
@@ -21,7 +21,7 @@ public class CustomPassageFormatterTests extends ESTestCase {
         String content = "This is a really cool highlighter. Unified highlighter gives nice snippets back. No matches here.";
 
-        CustomPassageFormatter passageFormatter = new CustomPassageFormatter("<em>", "</em>", new DefaultEncoder());
+        CustomPassageFormatter passageFormatter = new CustomPassageFormatter("<em>", "</em>", new DefaultEncoder(), 0);
```
Does it make sense to add context in the Test Class to this new argument used ?

yashdamani replied:
I could, but wouldn't going the class `CustomPassageFormatter` provide that context? I'm assuming you're suggesting adding context that the new parameter is for `numberOfFragments`? Correct me if I am wrong in my understanding your suggestion.

yashdamani commented:
@mayya-sharipova would you be able to help with reviewing this PR?

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova reviewed (changes requested):
@yashdamani Thank you for your contribution. The approach makes sense to me.


Please add rest tests into `rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.highlight` for the change. Here is the guidance about [YAML rest tests](https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#testing-the-rest-layer).

yashdamani committed:
changed datatype to int for numberOfFragments and added valid keyword for changelog file

yashdamani committed:
added rest tests for PR #106306

yashdamani commented:
@mayya-sharipova I've added the rest test file for the change. I'm not sure the about preceding number in the naming convention for the .yml files. I've added the case for number_of_fragments > 0 and number_of_fragments = 0. Is a default case (where the param number_of_fragments isn't mentioned) also required? Also, I've used the same input as the original issue linked with this PR.

mayya-sharipova commented:
@elasticsearchmachine test this please

mayya-sharipova commented on docs/changelog/106306.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 99961
+summary: "added fix for inconsistent text trimming in Unified Highlighter"
+area: Search
```
please change to :  `area: Highlighting`

yashdamani commented:
@mayya-sharipova I ran `./gradlew check` earlier in the day, and it ran for almost an hour, and I ended up exiting it. Is that expected?

yashdamani committed:
changed area to Highlighting in changelog and added a change to skip this test for nodes than run version < 8.14

yashdamani commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/510_fragment_trimming_fix.yml:
```diff
@@ -0,0 +1,59 @@
+setup:
```
Thank you! Made all the changes as suggested.

mayya-sharipova commented:
@yashdamani Indeed `./gradlew check` runs long, we usually don't run it locally. We run `./gradlew precommit` and then applicable tests.

mayya-sharipova commented:
@elasticsearchmachine test this please

mayya-sharipova commented:
@yashdamani Are you are able to access CI build failures and see and correct tests that are failing? 

```
 ./gradlew ':server:test' --tests "org.elasticsearch.lucene.search.uhighlight.CustomUnifiedHighlighterTests.testRepeatPhrase" -Dtests.seed=512E5E4D03C679D1 -Dtests.locale=de-LU -Dtests.timezone=America/Argentina/Buenos_Aires -Druntime.java=21

./gradlew ':server:test' --tests "org.elasticsearch.lucene.search.uhighlight.CustomUnifiedHighlighterTests.testSmallSentenceBoundedBreakIterator" -Dtests.seed=512E5E4D03C679D1 -Dtests.locale=de-LU -Dtests.timezone=America/Argentina/Buenos_Aires -Druntime.java=21
 

<br class="Apple-interchange-newline">
```

yashdamani commented:
@mayya-sharipova I can't access the builds. On clicking Details it asks me to sign in to buildkite.

mayya-sharipova commented:
@yashdamani The test failures are only in `CustomUnifiedHighlighterTests`, so you can run:

 ` ./gradlew ':server:test' --tests "org.elasticsearch.lucene.search.uhighlight.CustomUnifiedHighlighterTests.*`

yashdamani committed:
added tests to CustomPassageFormatterTests to include different values of numberOfFragments

yashdamani reviewed (commented):
@mayya-sharipova there were errors in `CustomPassageFormatterTests` as well. I've made changes to now include tests for different values of `numberOfFragments`.  Would [this](https://github.com/elastic/elasticsearch/pull/106306/commits/646ca23fda9ef3e6cd01e791885d770d464932d0) be the right approach to modify the tests? If yes, then I'll use the same approach for `CustomUnifiedHighlighterTests` as well, which failed in the build.

mayya-sharipova commented:
@yashdamani Please don't force push changes while PR is under review, as it is difficult to see the changes from the previous review.

yashdamani committed:
added numberOfFragments param to tests

yashdamani committed:
added numberOfFragments param to CustomUnifiedHighlighterTests

yashdamani commented:
@mayya-sharipova alright, will keep that in mind going ahead.

yashdamani commented on server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java:
```diff
@@ -21,7 +21,10 @@ public class CustomPassageFormatterTests extends ESTestCase {
     public void testSimpleFormat() {
         String content = "This is a really cool highlighter. Unified highlighter gives nice snippets back. No matches here.";
 
-        CustomPassageFormatter passageFormatter = new CustomPassageFormatter("<em>", "</em>", new DefaultEncoder());
```
Got it. I was under the impression that both the testcases (where `numberOfFragments` is 0 and > 0) need to be written.

mayya-sharipova reviewed (approved):
@yashdamani Thanks for your great work! Merging for the 8.14 release.

yashdamani commented:
@mayya-sharipova Thank you!
